File: chapter03/python/merge_sort.py
class MergeSort:
    def __init__(self):
        self.nums = None

    def merge(self, nums, lo, mid, hi):
        i = lo
        j = mid + 1
        for k in range(lo, hi + 1):
            self.nums[k] = nums[k]

        for k in range(lo, hi + 1):
            if i > mid:
                nums[k] = self.nums[j]
                j += 1
            elif j > hi:
                nums[k] = self.nums[i]
                i += 1

            elif self.nums[i] > self.nums[j]:
                nums[k] = self.nums[j]
                j += 1
            else:
                nums[k] = self.nums[i]
                i += 1

    def _merge_sort(self, nums, lo, hi):
        if lo >= hi:
            return
        mid = (lo + hi) // 2
        self._merge_sort(nums, lo, mid)
        self._merge_sort(nums, mid + 1, hi)
        self.merge(nums, lo, mid, hi)

    def merge_sort(self, nums):
        self.nums = [0] * len(nums)
        self._merge_sort(nums, 0, len(nums) - 1)
        return nums


if __name__ == "__main__":
    nums = [3, 6, 1, 9, 7, 2, 4]
    ms = MergeSort()
    new_nums = ms.merge_sort(nums)
    print(new_nums)

File: chapter03/cpp/merge_sort.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <stdlib.h>

using namespace std;

void merge(vector<int>& nums, int lo, int mid, int hi)
{
    vector<int> tmp_nums(nums.begin(), nums.end());
    
    int i = lo, j = mid + 1;
    for(int k=lo; k<hi+1; k++)
    {   
        int e = tmp_nums[k];
        if(i > mid)
        {
            nums[k] = tmp_nums[j];
            j++;
        }
        else if(j > hi)
        {
            nums[k] = tmp_nums[i];
            i++;
        }
        else if(tmp_nums[i] > tmp_nums[j])
        {
            nums[k] = tmp_nums[j];
            j++;
        }
        else
        {
            nums[k] = tmp_nums[i];
            i++;
        }
    }
}


void merge_sort(vector<int>& nums, int lo, int hi)
{
    if(lo == hi)
        return;
    
    int mid = (lo + hi) / 2;
   
    merge_sort(nums, lo, mid);  
    merge_sort(nums, mid+1, hi);
    merge(nums, lo, mid, hi);
}

int main()
{
    vector<int> nums = {3, 6, 1, 9, 7, 2, 4};
    merge_sort(nums, 0, nums.size()-1);
    for(auto& x: nums)
        cout << x << endl;
    return 0;
}